### PR TITLE
Added constructor so object initialize correctly on Windows. Fixed ou…

### PR DESCRIPTION
**Pull request by chaoticbob**
…t of bounds error when processing SPIR-V - this was throwing a vector out of bounds exception on some Window's (MSVC) build variations.

Added bash based tester. Changes were tested on both Linux and Windows to make sure existing functionality remained in tact.

### CODE REVIEW

File: spir2cross.cpp [hunk 1]
```diff
@@ -33,10 +33,17 @@ Instruction::Instruction(const vector<uint32_t> &spirv, uint32_t &index)
...
-
-    if (index > spirv.size())
-        throw CompilerError("SPIR-V instruction goes out of bounds.");
+	index += count;
+
+	if (spirv.size() == index) {
+		offset = UINT32_MAX;
+		index = UINT32_MAX;
+	}
+	else {
```
Comment (HansKristian-ARM) [inline spir2cross.cpp]: else if?
Also, I would prefer if the bracing is consistent (Allman, not K&R).
I will consider adding clang-format after this :)

File: spir2cross.cpp [hunk 2]
```diff
@@ -463,11 +470,19 @@ void Compiler::parse()
     meta.resize(bound);
 
     uint32_t offset = 5;
-    while (offset < len)
-        inst.emplace_back(spirv, offset);
-
-    for (auto &i : inst)
+    bool hr = offset < len;
+    while (hr) {
+        Instruction i = Instruction( spirv, offset );
```
Comment (HansKristian-ARM) [inline spir2cross.cpp]: emplace_back?

File: spir2cross.hpp [hunk 3]
```diff
@@ -225,7 +225,7 @@ namespace spir2cross
                     return nullptr;
             }
 
-            struct
+            struct execution
```
Comment (HansKristian-ARM) [inline spir2cross.hpp]: Upper-case for Execution and Source.

File: spir2cross.hpp [hunk 4]
```diff
@@ -236,13 +236,15 @@ namespace spir2cross
                 } workgroup_size;
                 uint32_t invocations = 0;
                 uint32_t output_vertices = 0;
+				execution() {}
```
Comment (HansKristian-ARM) [inline spir2cross.hpp]: Can this be = default; ?
This seems like a buggy MSVC version that is causing this, so if it has to be {} that is fine as well, but I'd like a comment about this workaround.

File: spir2cross.hpp [hunk 5]
```diff
@@ -236,13 +236,15 @@ namespace spir2cross
...
             } execution;
 
-            struct
+            struct source
             {
                 uint32_t version = 0;
                 bool es = false;
                 bool known = false;
+				source() {}
```
Comment (HansKristian-ARM) [inline spir2cross.hpp]: Ditto.

File: test_shaders.bash [hunk 6]
```diff
@@ -0,0 +1,83 @@
...
+    src_dir=$THIS_DIR/shaders/$dir
+    src_shaders=$(ls -C1 $src_dir/*)
+
+    spirv_dir=$TMP_DIR/spirv/$dir
+    mkdir -p $spirv_dir
+
+    for src_file in ${src_shaders[*]}
+    do
+        spirv_file=$spirv_dir/`basename $src_file`.spirv
+        glslangvalidator -V -o $spirv_file $src_file
```
Comment (HansKristian-ARM) [inline test_shaders.bash]: Typo in glslangValidator. Windows? :)

File: test_shaders.bash [hunk 7]
```diff
@@ -0,0 +1,83 @@
...
+    mkdir -p $out_dir
+
+    for src_file in ${src_shaders[*]}
+    do
+        basename=`basename $src_file`;
+        spirv_file=$spirv_dir/$basename.spirv
+        glsl_file=$out_dir/$basename
+        if [ -f "$spirv_file" ]; then
+            echo "spir2cross --output $glsl_file $spirv_file"
+            spir2cross --output $glsl_file $spirv_file
```
Comment (HansKristian-ARM) [inline test_shaders.bash]: ./spir2cross for Unix, . is not in default $PATH.

File: test_shaders.bash [hunk 8]
```diff
@@ -0,0 +1,83 @@
...
+        fi
+    done
+done
+
+echo ""
+
+# Diff report
+echo "This is a diff of the shader files that compiled succesffully to SPIR_V and back to GLSL." > $REPORT_FILE
```
Comment (HansKristian-ARM) [inline test_shaders.bash]: succesffully typo

File: spir2cross.cpp [hunk 9]
```diff
@@ -774,12 +789,13 @@ void Compiler::unset_decoration(uint32_t id, Decoration decoration)
 
 void Compiler::parse(const Instruction &i)
 {
-    auto ops = stream(i.offset);
+    auto ops = (UINT32_MAX == i.offset) ? nullptr : stream(i.offset);
     auto op = static_cast<Op>(i.op);
     uint32_t length = i.length;
 
-    if (i.offset + length > spirv.size())
+    if ((UINT32_MAX != i.offset) && (i.offset + length > spirv.size())) {
```
Comment (HansKristian-ARM) [inline spir2cross.cpp]: i.offset == UINT32_MAX.

File: spir2cross.cpp [hunk 10]
```diff
@@ -463,11 +470,19 @@ void Compiler::parse()
...
-    while (offset < len)
-        inst.emplace_back(spirv, offset);
-
-    for (auto &i : inst)
+    bool hr = offset < len;
+    while (hr) {
+        Instruction i = Instruction( spirv, offset );
+		inst.push_back( i );
+		if(UINT32_MAX == offset) {
+            hr = false;
```
Comment (HansKristian-ARM) [inline spir2cross.cpp]: Formatting is inconsistent here.
It's very little code so I can probably fix it up myself (and add clang-format) after review.

**Comment by HansKristian-ARM** [comment]
I'm a bit unsure about adding another test script that duplicates test_shaders.py. Is the main motivation for having a bash script that it's harder to run Python than Bash on Windows? If it's the malisc dependency that is a problem, that can easily be made optional/removed.

**Comment by HansKristian-ARM** [comment]
Sorry for long delay, this project has been adopted by Khronos, and is now hosted here: https://github.com/KhronosGroup/SPIRV-Cross. This repository will be disabled shortly. I believe I have fixed this issue correctly in the Khronos repository. If the issue persists, please re-open a pull request there.
